File: PSME/application-ref/src/rest-ref/constants/storage.cpp

/*!
 * @copyright
 * Copyright (c) 2015-2016 Intel Corporation
 *
 * @copyright
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * @copyright
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * @copyright
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * */

#include "psme/rest-ref/constants/storage.hpp"

namespace app {
namespace rest {
namespace constants {

namespace LogicalDrive {
     const char TYPE[] = "Type";
     const char MODE[] = "Mode";
     const char PROTECTED[] = "Protected";
     const char SNAPSHOT[] = "Snapshot";
     const char CAPACITY[] = "CapacityGiB";
     const char IMAGE[] = "Image";
     const char BOOTABLE[] = "Bootable";
     const char LOGICAL_DRIVES[] = "LogicalDrives";
     const char PHYSICAL_DRIVES[] = "PhysicalDrives";
     const char MASTER_DRIVE[] = "MasterDrive";
     const char TARGETS[] = "Targets";
     const char CLONE[] = "Clone";
     const char TARGET[] = "target";
}

namespace PhysicalDrive {
}

namespace StorageService {
     const char TARGETS[] = "Targets";
     const char REMOTE_TARGETS[] = "RemoteTargets";
     const char LOGICAL_DRIVES[] = "LogicalDrives";
     const char DRIVES[] = "Drives";
}

namespace RemoteTarget {
     const char TYPE[] = "Type";
     const char ADDRESSES[] = "Addresses";
     const char ISCSI[] = "iSCSI";
     const char TARGET_LUN[] = "TargetLUN";
     const char LUN[] = "LUN";
     const char LOGICAL_DRIVE[] = "LogicalDrive";
     const char TARGET_IQN[] = "TargetIQN";
     const char TARGET_PORTAL_IP[] = "TargetPortalIP";
     const char TARGET_PORTAL_PORT[] = "TargetPortalPort";
     const char INITIATOR[] = "Initiator";
     const char INITIATOR_IQN[] = "InitiatorIQN";
     const char LOGICAL_DRIVES[] = "LogicalDrives";
}

}
}
}

File: PODM/SW/node-composer/src/main/java/com/intel/rsd/nodecomposer/persistence/redfish/Storage.java
/*
 * Copyright (c) 2016-2019 Intel Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.intel.rsd.nodecomposer.persistence.redfish;

import com.intel.rsd.nodecomposer.persistence.redfish.base.DiscoverableEntity;

import javax.persistence.JoinColumn;
import javax.persistence.ManyToOne;
import javax.persistence.OneToMany;
import javax.persistence.Table;
import java.util.HashSet;
import java.util.Objects;
import java.util.Set;

import static com.intel.rsd.nodecomposer.utils.Contracts.requiresNonNull;
import static javax.persistence.CascadeType.MERGE;
import static javax.persistence.CascadeType.PERSIST;
import static javax.persistence.FetchType.LAZY;

@javax.persistence.Entity
@Table(name = "storage")
public class Storage extends DiscoverableEntity {
    @OneToMany(mappedBy = "storage", fetch = LAZY, cascade = {MERGE, PERSIST})
    private Set<Drive> drives = new HashSet<>();

    @ManyToOne(fetch = LAZY, cascade = {MERGE, PERSIST})
    @JoinColumn(name = "computer_system_id")
    private ComputerSystem computerSystem;

    @ManyToOne(fetch = LAZY, cascade = {MERGE, PERSIST})
    @JoinColumn(name = "chassis_id")
    private Chassis chassis;

    public Set<Drive> getDrives() {
        return drives;
    }

    public void addDrive(Drive drive) {
        requiresNonNull(drive, "drive");

        drives.add(drive);
        if (!this.equals(drive.getStorage())) {
            drive.setStorage(this);
        }

        if (chassis != null) {
            drive.setChassis(chassis);
        }
    }

    public void unlinkDrive(Drive drive) {
        if (drives.contains(drive)) {
            drives.remove(drive);
            if (drive != null) {
                drive.unlinkStorage(this);
                drive.setChassis(null);
            }
        }
    }

    public ComputerSystem getComputerSystem() {
        return computerSystem;
    }

    public void setComputerSystem(ComputerSystem computerSystem) {
        if (!Objects.equals(this.computerSystem, computerSystem)) {
            unlinkComputerSystem(this.computerSystem);
            this.computerSystem = computerSystem;
            if (computerSystem != null && !computerSystem.getStorages().contains(this)) {
                computerSystem.addStorage(this);
            }
        }
    }

    public void unlinkComputerSystem(ComputerSystem computerSystem) {
        if (Objects.equals(this.computerSystem, computerSystem)) {
            this.computerSystem = null;
            if (computerSystem != null) {
                computerSystem.unlinkStorage(this);
            }
        }
    }

    public Chassis getChassis() {
        return chassis;
    }

    public void setChassis(Chassis chassis) {
        if (!Objects.equals(this.chassis, chassis)) {
            unlinkChassis(this.chassis);
            this.chassis = chassis;
            if (chassis != null && !chassis.getStorages().contains(this)) {
                chassis.addStorage(this);
            }
        }
    }

    public void unlinkChassis(Chassis chassis) {
        if (Objects.equals(this.chassis, chassis)) {
            this.chassis = null;
            if (chassis != null) {
                chassis.unlinkStorage(this);
            }
        }
    }

    @Override
    public void preRemove() {
        unlinkCollection(drives, this::unlinkDrive);
        unlinkComputerSystem(computerSystem);
        unlinkChassis(chassis);
    }
}
